nav2_costmap_2d/src/costmap_2d_ros.cpp

```diff
@@ -308,6 +331,10 @@ Costmap2DROS::on_cleanup(const rclcpp_lifecycle::State & /*state*/)
   costmap_publisher_.reset();
   clear_costmap_service_.reset();
 
+  for (auto & layer_pub : layer_publishers_) {
+    layer_pub.reset();
```

Comment:
I think the publishers need to be reset before `layered_costmap_.reset();`. That will reset the layers which the publishers have a raw pointer to which would become null. It wouldn't be an issue for most situations, but there is a service in the costmap publisher that someone could try to invoke causing a crash at this point. Basically just move L331 and this for loop before that line.